external/storm-kafka-new-consumer-api/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java

```diff
@@ -0,0 +1,445 @@
+        initialized = false;
+
+        // Spout internals
+        this.collector = collector;
+        maxRetries = kafkaSpoutConfig.getMaxTupleRetries();
+
+        // Offset management
+        firstPollOffsetStrategy = kafkaSpoutConfig.getFirstPollOffsetStrategy();
+        consumerAutoCommitMode = kafkaSpoutConfig.isConsumerAutoCommitMode();
+
+        if (!consumerAutoCommitMode) {     // If it is auto commit, no need to commit offsets manually
+            timer = new Timer(kafkaSpoutConfig.getOffsetsCommitFreqMs(), 500, TimeUnit.MILLISECONDS);
+            acked = new HashMap<>();
+        }
+
+        // Kafka consumer
+        kafkaConsumer = new KafkaConsumer<>(kafkaSpoutConfig.getKafkaProps(),
+                kafkaSpoutConfig.getKeyDeserializer(), kafkaSpoutConfig.getValueDeserializer());
+        kafkaConsumer.subscribe(kafkaSpoutConfig.getSubscribedTopics(), new KafkaSpoutConsumerRebalanceListener());
+        // Initial poll to get the consumer registration process going.
+        // KafkaSpoutConsumerRebalanceListener will be called foloowing this poll upon partition registration
+        kafkaConsumer.poll(0);
+
+        LOG.debug("Kafka Spout opened with the following configuration: {}", kafkaSpoutConfig.toString());
+    }
+
+    // ======== Next Tuple =======
+
+    @Override
+    public void nextTuple() {
+        if (initialized) {
+            if(commit()) {
+                commitOffsetsForAckedTuples();
+            } else {
+                emitTuples(poll());
+            }
+        } else {
+            LOG.debug("Spout not initialized. Not sending tuples until initialization completes");
+        }
+    }
+
+    private boolean commit() {
+        return !consumerAutoCommitMode && timer.expired();    // timer != null for non auto commit mode
+    }
+
+    private ConsumerRecords<K, V> poll() {
+        final ConsumerRecords<K, V> consumerRecords = kafkaConsumer.poll(kafkaSpoutConfig.getPollTimeoutMs());
+        LOG.debug("Polled [{}] records from Kafka", consumerRecords.count());
+        return consumerRecords;
+    }
+
+    private void emitTuples(ConsumerRecords<K, V> consumerRecords) {
+        for (TopicPartition tp : consumerRecords.partitions()) {
+            final Iterable<ConsumerRecord<K, V>> records = consumerRecords.records(tp.topic());     // TODO Decide if want to give flexibility to emmit/poll either per topic or per partition
+
+            for (final ConsumerRecord<K, V> record : records) {
+                if (record.offset() == 0 || record.offset() > acked.get(tp).committedOffset) {      // The first poll includes the last committed offset. This if avoids duplication
+                    final List<Object> tuple = tupleBuilder.buildTuple(record);
+                    final MessageId messageId = new MessageId(record, tuple);                                  // TODO don't create message for non acking mode. Should we support non acking mode?
```

Comment:
In non acking mode ack is called instantly.

https://github.com/apache/storm/blob/master/storm-core/src/clj/org/apache/storm/daemon/executor.clj#L592-L595

So we already support it.  Please remove the comment.